api/app/utils/geospatial.py

```diff
@@ -0,0 +1,165 @@
+import os
+from dataclasses import dataclass
+import logging
+from typing import Any, Optional
+from aiobotocore.client import AioBaseClient
+from botocore.exceptions import ClientError
+from osgeo import gdal
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Vector2RasterOptions:
+    """
+    Inputs required for burning vector geospatial data into a raster
+    """
+
+    source_geotransform: Any
+    source_projection: Any
+    source_x_size: int
+    source_y_size: int
+
+
+@dataclass
+class GeospatialOptions:
+    """
+    Options supplied by caller to declare optional outputs
+    """
+
+    vector_options: Optional[Vector2RasterOptions] = None
+    include_geotransform: Optional[bool] = True
+    include_projection: Optional[bool] = True
+    include_x_size: Optional[bool] = None
+    include_y_size: Optional[bool] = None
+
+
+@dataclass
+class GeospatialReadResult:
+    """
+    Raster data array and caller requested metadata
+    """
+
+    data_array: Any
+    data_geotransform: Optional[Any]
+    data_projection: Optional[Any]
+    data_x_size: Optional[int]
+    data_y_size: Optional[int]
+
+
+def get_geospatial_metadata(data_source, options: Optional[GeospatialOptions]):
+    """
+    Extracts metadata defined by the options parameter from the data_source raster
+
+    :param data_source: input raster to extract metadata from
+    :param options: defines the metadata the caller is interested in extracting from the raster
+    :return: optional metadata of geotransform, project, x and y sizes from the raster
+    """
+    geotransform = None
+    projection = None
+    x_size = None
+    y_size = None
+    if options is None:
+        return (geotransform, projection.x_size, y_size)
+
+    if options.include_geotransform:
+        geotransform = data_source.GetGeoTransform()
+
+    if options.include_projection:
+        projection = data_source.GetProjection()
+
+    if options.include_x_size:
+        x_size = data_source.RasterXSize
+
+    if options.include_y_size:
+        y_size = data_source.RasterYSize
+
+    return (geotransform, projection, x_size, y_size)
+
+
+def read_raster_data(data_source):
+    """
+    Read raster data and return as array from the data source
+
+    :param data_source: data source to read the data from
+    :return: raster data as array
+    """
+    data_band = data_source.GetRasterBand(1)
```

Comment:
Some of the grib files we work with (NAM and GFS I think) have all the weather parameters in a single file, each in their own band. We could consider adding an optional `band` parameter to this function with a default of 1.
`def read_raster_data(data_source, band=1)`

Reply:
Nice, done in: https://github.com/bcgov/wps/pull/3800/commits/615e2b1bb8966d2ca1986253abfe559532bf5796